File: src/main/java/com/rg/webmvctest/adapters/memoryrepository/InMemoryUserRegistrationRepository.java
package com.rg.webmvctest.adapters.memoryrepository;

import com.rg.webmvctest.domain.DuplicateRegistrationException;
import com.rg.webmvctest.domain.RegistrationNotFoundException;
import com.rg.webmvctest.domain.UserRegistration;
import com.rg.webmvctest.domain.UserRegistrationRepository;
import kotlin.text.Charsets;
import org.springframework.stereotype.Component;
import org.springframework.util.DigestUtils;

import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

@Component
public class InMemoryUserRegistrationRepository implements UserRegistrationRepository {

    private final SecureRandom secureRandom = new SecureRandom();
    private final ConcurrentHashMap<String, UserRegistration> storage = new ConcurrentHashMap<>();


    @Override
    public UserRegistration findById(String registrationId) {
        return storage.get(registrationId);
    }

    @Override
    public UserRegistration getById(String registrationId) {
        if (!storage.containsKey(registrationId)) {
            throw new RegistrationNotFoundException(registrationId);
        }

        return storage.get(registrationId);
    }

    @Override
    public UserRegistration insert(String emailAddress, String name, String lastName) {
        if (findByEmailAddress(emailAddress) != null) {
            throw new DuplicateRegistrationException(emailAddress);
        }

        UserRegistration registration = new UserRegistration(
                generateId(emailAddress),
                emailAddress,
                name,
                lastName
        );

        storage.put(registration.getRegistrationId(), registration);
        return registration;
    }

    @Override
    public UserRegistration update(UserRegistration userRegistration) {
        if (!storage.containsKey(userRegistration.getRegistrationId())) {
            throw new RegistrationNotFoundException(userRegistration.getRegistrationId());
        }

        storage.put(userRegistration.getRegistrationId(), userRegistration);
        return userRegistration;
    }

    @Override
    public UserRegistration findByEmailAddress(String emailAddress) {
        return storage.values().stream().filter(it -> it.getEmailAddress().equals(emailAddress)).findFirst().orElse(null);
    }

    @Override
    public List<UserRegistration> findAll() {
        return new ArrayList<>(storage.values());
    }

    /**
     * This is an UNSAFE method to produce any sort of ID since MD5 is unsafe.
     * It is used here only to create short text ids.
     */
    private String generateId(String emailAddress) {
        int salt = secureRandom.nextInt();
        return DigestUtils.md5DigestAsHex(String.format("%d%s", salt, emailAddress).getBytes(Charsets.UTF_8)).substring(0, 8);
    }
}

File: src/main/kotlin/com/rg/webmvctest/adapters/memoryrepository/InMemoryUserRegistrationRepository.kt
package com.rg.webmvctest.adapters.memoryrepository

import com.rg.webmvctest.domain.DuplicateRegistrationException
import com.rg.webmvctest.domain.RegistrationNotFound
import com.rg.webmvctest.domain.UserRegistration
import com.rg.webmvctest.domain.UserRegistrationRepository
import org.springframework.stereotype.Component
import org.springframework.util.DigestUtils
import java.security.SecureRandom
import java.util.concurrent.ConcurrentHashMap

@Component
open class InMemoryUserRegistrationRepository : UserRegistrationRepository {

    private val secureRandom = SecureRandom()
    private val storage = ConcurrentHashMap<String, UserRegistration>()

    override fun findById(registrationId: String): UserRegistration? =
            storage[registrationId]

    override fun getById(registrationId: String): UserRegistration =
            storage[registrationId] ?: throw RegistrationNotFound(registrationId)

    override fun insert(emailAddress: String, name: String, lastName: String): UserRegistration {
        if (findByEmailAddress(emailAddress) != null) {
            throw DuplicateRegistrationException(emailAddress)
        }

        val registration = UserRegistration(
                registrationId = generateId(emailAddress),
                emailAddress = emailAddress,
                name = name,
                lastName = lastName
        )
        storage[registration.registrationId] = registration

        return registration
    }

    override fun update(userRegistration: UserRegistration): UserRegistration {
        if (storage[userRegistration.registrationId] == null) {
            throw RegistrationNotFound(userRegistration.registrationId)
        }

        storage[userRegistration.registrationId] = userRegistration
        return userRegistration
    }

    override fun findByEmailAddress(emailAddress: String): UserRegistration? =
            storage.values.firstOrNull { it.emailAddress == emailAddress }

    override fun findAll(): List<UserRegistration> = storage.values.toList()

    /**
     * This is an UNSAFE method to produce any sort of ID since MD5 is unsafe.
     * It is used here only to create short text ids.
     */
    private fun generateId(emailAddress: String): String =
         DigestUtils.md5DigestAsHex("${secureRandom.nextInt()}$emailAddress".toByteArray(Charsets.UTF_8)).take(8)
}